src/vmm/src/devices/virtio/net/device.rs

```diff
@@ -154,18 +157,37 @@ impl RxBuffers {
     /// * The `RxBuffers` should include at least one parsed `DescriptorChain`.
     /// * `size` needs to be smaller or equal to total length of the first `DescriptorChain` stored
     ///   in the `RxBuffers`.
-    unsafe fn mark_used(&mut self, size: u32) -> ParsedDescriptorChain {
-        // Since we were able to write a frame in guest memory, we should have at least one
-        // descriptor chain here. If not, we have a bug, so fail fast, since the device is
-        // fundamentally broken.
-        let mut parsed_dc = self.parsed_descriptors.pop_front().expect(
-            "net: internal bug. Mismatch between written frame size and available descriptors",
-        );
+    unsafe fn mark_used(&mut self, mut bytes_written: u32, rx_queue: &mut Queue) {
+        self.used_bytes = bytes_written;
+
+        let mut used_heads: u16 = 0;
+        for parsed_dc in self.parsed_descriptors.iter() {
+            let used_bytes = bytes_written.min(parsed_dc.length);
+            // Safe because we know head_index isn't out of bounds
+            rx_queue
+                .write_used_element(
+                    (rx_queue.next_used + Wrapping(used_heads)).0,
```

Comment:
I would prefer it if we only passed `used_heads` to `write_used_element` and the calculation happens inside there. Like that, we don't need to expose `Queue` internals here.

Reply:
done